2025-Robot-Code/src/main/java/frc/robot/commands/SelfDriveCommands.java

```diff
@@ -0,0 +1,341 @@
+	private static class ShiftingApproachTarget {
+		private final Pose2d targetPose;
+		private final Pose2d startFrom;
+		private final double initialDistance;
+		private boolean hasBeenNearStart = false;
+
+		public ShiftingApproachTarget(Pose2d actual, Pose2d startShifting, Pose2d target) {
+			/**
+			 * The initial pose of the robot when the command started. This can be used
+			 * to determine how much closer
+			 */
+			targetPose = target;
+			if (getDistance(actual, target) < getDistance(startShifting, target)) {
+				var proj = actual.relativeTo(target).getTranslation().times(0.1);
+				startFrom = new Pose2d(proj, startShifting.getRotation());
+			} else
+				startFrom = startShifting;
+			initialDistance = getDistance(actual, targetPose);
+		}
+
+		/**
+		 * Determines whether or not the robot is near the starting point of this path
+		 * 
+		 * @param actual The actual pose of the robot
+		 * @return
+		 */
+		public boolean isNearStartpoint(Pose2d actual) {
+			return MathUtil.isNear(0, getDistance(actual, startFrom), 1);
+		}
+
+		public Pose2d getTarget(Pose2d current) {
+			double newDistance = getDistance(current, targetPose);
+
+			// ensure that we are close enough to the starting point before starting the
+			// curved approach
+			if (!isNearStartpoint(current) && !hasBeenNearStart)
+				return startFrom;
+			else
+				hasBeenNearStart = true;
+
+			// this(newDist / initialDistance) should NEVER be greater than 1
+			double progress = 1 - MathUtil.clamp(newDistance / initialDistance, 0, 1);
+
+			// we get the progression from startTarget -> endTarget, adding 0.05 to make
+			// sure if the robot was exactly in the right position it doesn't stop the robot
+			double clamped = MathUtil.clamp(progress + 0.05, 0, 1);
+			Translation2d interpolatedTranslation = startFrom.getTranslation().interpolate(
+					targetPose.getTranslation(),
+					clamped);
+			Rotation2d interpolatedRotation = startFrom.getRotation().interpolate(targetPose.getRotation(),
+					clamped);
+
+			var target = new Pose2d(interpolatedTranslation, interpolatedRotation);
+
+			// TODO: tune this constant to get the greatest and most ACCURATE alignment to
+			// the target while allowing for the greatest distance
+			if (progress < 0.5) {
+				// shift the target to the pose if we aren't halfway there (makes kind of an
+				// J-shape path)
+				// TODO: tune the amount we shift the pose by so that the PID is very agressive
+				var shift = new Transform2d(target, current).times(0.1).times(1 - (progress / 0.5));
+
+				target = target.plus(shift);
+			}
+
+			return target;
+		}
+	}
+
+	private static Command reefApproach(
+			Pose2d scoringPose,
+			Drivetrain drive,
+			boolean preferTopInRepositioning,
+			Pose2d startFrom,
+			DoubleSupplier driverX,
+			DoubleSupplier driverY,
+			DoubleSupplier driverRotX,
+			DoubleSupplier driverRotY,
+			DoubleConsumer updateProgress) {
+		final Supplier<Pose2d> robotPose = () -> drive.getState().Pose;
```

Comment:
why are we creating this supplier here, instead of pulling it from the function call?